iothub/device/src/InternalClient.cs

```diff
@@ -474,65 +371,120 @@ public async Task UpdateReportedPropertiesAsync(TwinCollection reportedPropertie
+            object userContext,
+            CancellationToken cancellationToken = default)
         {
-            Argument.AssertNotNullOrWhiteSpace(lockToken, nameof(lockToken));
+            if (Logging.IsEnabled)
+                Logging.Enter(this, methodName, methodHandler, userContext, nameof(SetMethodHandlerAsync));
+
+            Argument.AssertNotNullOrWhiteSpace(methodName, nameof(methodName));
             cancellationToken.ThrowIfCancellationRequested();
 
+            await _methodsSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
             try
             {
-                await InnerHandler.CompleteMessageAsync(lockToken, cancellationToken).ConfigureAwait(false);
+                if (methodHandler != null)
+                {
+                    await HandleMethodEnableAsync(cancellationToken).ConfigureAwait(false);
+                    _deviceMethods[methodName] = new Tuple<Func<DirectMethodRequest, object, Task<DirectMethodResponse>>, object>(methodHandler, userContext);
+                }
+                else
+                {
+                    _deviceMethods.Remove(methodName);
+                    await HandleMethodDisableAsync(cancellationToken).ConfigureAwait(false);
+                }
             }
             catch (IotHubClientException ex) when (ex.InnerException is OperationCanceledException)
             {
                 cancellationToken.ThrowIfCancellationRequested();
                 throw;
             }
+            finally
+            {
+                _methodsSemaphore.Release();
+
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, methodName, methodHandler, userContext, nameof(SetMethodHandlerAsync));
+            }
         }
 
         /// <summary>
-        /// Deletes a received message from the device queue
+        /// Sets a new delegate that is called for a method that doesn't have a delegate registered for its name.
+        /// If a default delegate is already registered it will replace with the new delegate.
+        /// A method handler can be unset by passing a null MethodCallback.
         /// </summary>
-        /// <param name="message">The message to complete</param>
-        /// <param name="cancellationToken">A token to cancel the operation. </param>
-        /// <returns>The previously received message</returns>
-        public async Task CompleteMessageAsync(Message message, CancellationToken cancellationToken = default)
+        /// <param name="methodHandler">The delegate to be used when a method is called by the cloud service and there is no
+        /// delegate registered for that method name.</param>
+        /// <param name="userContext">Generic parameter to be interpreted by the client code.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice
+        /// of cancellation.</param>
+        public async Task SetMethodDefaultHandlerAsync(
```

Comment:
I had reordered these previously and aimed to put them in the rough order I thought someone would need to call them. IOW, setting handlers would come before telemetry or completing C2D messages etc.

You have a different order in mind. Can you elaborate on that?

Reply:
Ah, I see your point. The order I used was simply based on the ordering of operations that I see in our docs (which I feel could also corelate to increasing order of complexity of operations) - send event, receive event, direct method, twins. I can see the value in placing the callback APIs before the "sending" APIs, but I'd also like to have the same or similar operations grouped together, i.e. all twin operations grouped together. How do you feel about that? 
Also, with the inheritance model, since we'll be removing a bunch of APIs from either the InternalClient or (Device/Module)Client we can do the re-ordering again. The idea here was to make understanding the next PR's diff a bit better.